src/grad.jl

```diff
@@ -1,96 +1,46 @@
-export grad, jacobian, jvp, j′vp, to_vec
-replace_arg(x, xs::Tuple, k::Int) = ntuple(p -> p == k ? x : xs[p], length(xs))
-
 """
-    grad(fdm, f, xs...)
+    jacobian(fdm, f, x...)
 
-Approximate the gradient of `f` at `xs...` using `fdm`. Assumes that `f(xs...)` is scalar.
+Approximate the Jacobian of `f` at `x` using `fdm`. Results will be returned as a
+`Matrix{<:Number}` of `size(length(y_vec), length(x_vec))` where `x_vec` is the flattened
+version of `x`, and `y_vec` the flattened version of `f(x...)`. Flattening performed by
+[`to_vec`](@ref).
 """
-function grad end
-
-function _grad(fdm, f, x::AbstractArray{T}) where T <: Number
-    # x must be mutable, we will mutate it and then mutate it back.
-    dx = similar(x)
-    for k in eachindex(x)
-        dx[k] = fdm(zero(T)) do ϵ
-            xk = x[k]
-            x[k] = xk +  ϵ
-            ret = f(x)
-            x[k] = xk  # Can't do `x[k] -= ϵ` as floating-point math is not associative
+function jacobian(fdm, f, x::Vector{<:Number}; len=nothing)
+    len !== nothing && Base.depwarn("len parameter to jacobian is deprecated", :jacobian)
```

Comment:
I would tweak the wording a bit. (which unfortunately puts it over the line limit)
```suggestion
    len !== nothing && Base.depwarn(
        "`len` keyword argument to `jacobian` is nonlonger required " *
        "and will not be permitted in the future.",
         :jacobian
    )
```

Other than that looks perfect.

Reply:
typo in suggestion: "non" -> "no "